Azure Batch runtime security improvements and minor refactor.

ryumbra opened the pull request:
1. Driver now uses authentication token to communicate with Azure Batch;
2. Driver now uses a Shared Access Signature to communicate with Azure Storage
3. org.apache.reef.runtime.azbatch.util package was refactored into
sub-packages to better organize a growing number of classes.

tyclintw commented on lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/AzureStorageUtil.java:
```diff
@@ -1,153 +0,0 @@
-/*
```
Suggest using "git mv" to keep the history in the future.

ryumbra replied:
good call. I was actually doing the refactoring using IntelliJ. I'll see if I can fix this.

ryumbra replied:
No idea why this is happening. I did "git reset HEAD~1 --soft", moved the files around and the "git mv". Still, it's showing up as delete + add. Strangely, all other files that I refactored are showed as having been moved.

tyclintw commented on lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/storage/AzureStorageUtil.java:
```diff
@@ -0,0 +1,150 @@
+public class AzureStorageUtil {
+  private static final Logger LOG = Logger.getLogger(AzureStorageUtil.class.getName());
+
+  private static final int BLOB_SAS_TOKEN_VALIDITY_MINUTES = 30;
```
Still not a huge fan of these being static.

ryumbra replied:
I could turn them into parameters with default values. How does that sound?

tyclintw commented on lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/storage/AzureStorageUtil.java:
```diff
@@ -0,0 +1,150 @@
+
+  private final ICloudBlobClientProvider cloudBlobClientProvider;
+
+  private final String azureStorageContainerName;
```
Would it be better to make this a parameter to the different functions? I find a utility class with an internal state variable to be a bit confusing.

ryumbra replied:
Hm, the disadvantage of that is that I'll need to litter all classes that use AzureStorageUtil with azureStorageContainerName internal state. 
Or maybe, AzureStorageUtil should be called, something like AzureStorageClient? Let me know if you feel strongly either way.

tyclintw commented on lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/storage/AzureStorageUtil.java:
```diff
@@ -0,0 +1,150 @@
+
+    try {
+      final CloudBlobClient cloudBlobClient = this.cloudBlobClientProvider.getCloudBlobClient();
+      final CloudBlobContainer container = cloudBlobClient.getContainerReference(this.azureStorageContainerName);
```
Call "container.createIfNotExists();" should be consistent with the other functions instead of assuming an order or calling.

ryumbra replied:
it turns out, createIfNotExists will error out when called on driver side. That's because the driver works with container SAS token, rather than account key. So I'll leave it as is. Unless, you feel strongly about refactoring AzureStorageUtil into two components -- one to be used by the client with account key and another one to be used by the Driver with SAS token.

tyclintw commented on lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/storage/AzureStorageUtil.java:
```diff
@@ -0,0 +1,150 @@
+    return policy;
+  }
+
+  private SharedAccessBlobPolicy getSharedAccessContainerPolicy() {
```
nit: Pretty much identical to getSharedAccessBlobPolicy. Consider creating a shared method.

ryumbra replied:
changed.

tyclintw commented on lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runtime/azbatch/util/storage/SharedAccessSignatureCloudBlobClientProvider.java:
```diff
@@ -0,0 +1,89 @@
+      throws IOException {
+    try {
+      final String uri = cloudBlob.getStorageUri().getPrimaryUri().toString();
+      return new URI(uri + "?" + this.azureStorageContainerSASToken);
```
This will probably work fine because we know the format of the PrimaryUri, but we should always try to use UriBuilder instead of string manipulation.

ryumbra replied:
good call!

ryumbra committed:
Azure Batch runtime security improvements and minor refactor

1. Driver now uses authentication token to communicate with Azure Batch
2. Driver now uses a Shared Access Signature to communicate with Azure Storage
3. org.apache.reef.runtime.azbatch.util package was refactored into sub-packages to better organize a growing number of classes.

ryumbra committed:
Addressing PR comments
